<a href="https://colab.research.google.com/github/SRARNAB7/HDS_5230_07_Arnab/blob/main/Week%2009/Week_09_Assignment_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NAME- ARNAB SUTAR
# ASSIGNMENT- WEEK 09
**Week 09 - Machine Learning with Scikit-learn**

1) Among the different classification models included in the Python notebook, which model had the best overall performance? Support your response by referencing appropriate evidence.

Model     Train     Test

Logistic  0.7333    0.718

Null 0.6467 0.608

Logistic_L1_C_1 0.732 0.716

Logistic_L1_C_01 0.726 0.706

Logistic_L1_C_10 0.7347 0.718

Logistic_L1_C_auto 0.7233 0.708

Logistic_SL1_C_auto 0.7307 0.714

RandomForest_noCV 0.9993 0.69

RandomForest_CV 0.9987 0.702

RandomForest_CV2 0.7273 0.702

Based on the results presented in the screenshot, the Logistic regression model with L1 penalty and C=10 (Logistic_L1_C_10) demonstrated the best overall performance. It achieved the highest test accuracy of 0.718, which is a key indicator of how well the model generalizes to new, unseen data. In contrast, the Random Forest (no CV) model had an almost perfect training accuracy of 0.9993, but its test accuracy dropped sharply to 0.686, suggesting significant overfitting.

The Logistic_L1_C_10 model showed a strong balance between training and testing performance, with a training accuracy of 0.7347 and a test accuracy of 0.718. This close alignment indicates that the model generalizes well without being overly fitted to the training data. Therefore, among all the classification models evaluated, Logistic_L1_C_10 proved to be the most reliable and generalizable.

In [ ]:
## Import Modules
import os
import sys
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
from patsy import dmatrices
from sklearn.metrics import confusion_matrix
import sklearn
from sklearn import datasets

In [ ]:
## Set default figure size to be larger
## this may only work in matplotlib 2.0+!
matplotlib.rcParams['figure.figsize'] = [10.0,6.0]
## Enable multiple outputs from jupyter cells
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [ ]:
## Get Version information
print(sys.version)
print("Pandas version: {0}".format(pd.__version__))
print("Matplotlib version: {0}".format(matplotlib.__version__))
print("Numpy version: {0}".format(np.__version__))
print("SciKitLearn version: {0}".format(sklearn.__version__))

3.11.11 (main, Dec  4 2024, 08:55:07) [GCC 11.4.0]
Pandas version: 2.2.2
Matplotlib version: 3.10.0
Numpy version: 2.0.2
SciKitLearn version: 1.6.1


Check the working directory Set the working directory to make paths easier :)

In [ ]:
# Working Directory
import os
print("My working directory:\n" + os.getcwd())
# Set Working Directory
os.chdir(".")
print("My new working directory:\n" + os.getcwd())

My working directory:
/content
My new working directory:
/content


Patient Mortality Dataset We will use a dataset with a binary outcome of mortality as a motivating example.

This is a dataset of patients demographics and disease status, with mortality indicated. The dataset is here:

data\healthcare\patientAnalyticFile.csv

In practice, you most likely would have created a dataset like this from multiple other files after cleaning, reshaping, and joining them.

You can generalize this setup to any situation with a binary outcome, such as estimating the probability of a customer filing a warranty claim, or the probability of a transaction being fraudulent.

We will first import this dataset and examine the potential variables to use in our classification algorithm.

In [ ]:
## Set print limits
pd.options.display.max_rows = 10
## Import Data
df_patient = \
 pd.read_csv('./PatientAnalyticFile.csv')
df_patient

PatientID DateOfBirth  Gender      Race  Myocardial_infarction  \
0              1  1962-02-27  female  hispanic                      0   
1              2  1959-08-18    male     white                      0   
2              3  1946-02-15  female     white                      0   
3              4  1979-07-27  female     white                      0   
4              5  1983-02-19  female  hispanic                      0   
...          ...         ...     ...       ...                    ...   
19995      19996  1997-12-19  female     other                      0   
19996      19997  1984-03-31  female     white                      0   
19997      19998  1993-07-04  female     white                      0   
19998      19999  1984-04-17    male     other                      0   
19999      20000  1966-05-14  female     white                      0   

       Congestive_heart_failure  Peripheral_vascular_disease  Stroke  \
0                             0                            0       0   
1                             0                            0       0   
2                             0                            0       0   
3                             0                            0       0   
4                             0                            0       0   
...                         ...                          ...     ...   
19995                         0                            0       0   
19996                         0                            0       0   
19997                         0                            0       0   
19998                         0                            0       0   
19999                         0                            0       0   

       Dementia  Pulmonary  ...  Metastatic_solid_tumour  HIV  Obesity  \
0             0          0  ...                        0    0        0   
1             0          0  ...                        0    0        0   
2             0          0  ...                        0    1        0   
3             0          1  ...                        0    0        0   
4             0          0  ...                        0    0        0   
...         ...        ...  ...                      ...  ...      ...   
19995         0          0  ...                        0    0        0   
19996         0          0  ...                        0    1        0   
19997         0          0  ...                        0    0        1   
19998         0          0  ...                        0    0        0   
19999         0          0  ...                        0    0        0   

       Depression  Hypertension  Drugs  Alcohol  First_Appointment_Date  \
0               0             0      0        0              2013-04-27   
1               0             1      0        0              2005-11-30   
2               0             1      0        0              2011-11-05   
3               0             0      0        0              2010-03-01   
4               0             1      0        0              2006-09-22   
...           ...           ...    ...      ...                     ...   
19995           0             0      0        0              2008-06-14   
19996           0             1      0        0              2007-04-24   
19997           0             1      0        0              2010-10-16   
19998           0             1      0        0              2015-01-04   
19999           0             0      1        0              2011-04-01   

       Last_Appointment_Date  DateOfDeath  
0                 2018-06-01          NaN  
1                 2008-11-02   2008-11-02  
2                 2015-11-13          NaN  
3                 2016-01-17   2016-01-17  
4                 2018-06-01          NaN  
...                      ...          ...  
19995             2018-06-01          NaN  
19996             2018-06-01          NaN  
19997             2018-06-01          NaN  
19998             2018-06-01       

We need to make a variable to indicate mortality. We can do that based on the abscence of 'date of death':

In [ ]:
# Create mortality variable
df_patient['mortality'] = \
    np.where(df_patient['DateOfDeath'].isnull(),
             0,1)
# Examine
df_patient['mortality']

0        0
1        1
2        0
3        1
4        0
        ..
19995    0
19996    0
19997    0
19998    0
19999    0
Name: mortality, Length: 20000, dtype: int64

In [ ]:
df_patient['mortality'].describe()

count    20000.000000
mean         0.354700
std          0.478434
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          1.000000
Name: mortality, dtype: float64

In [ ]:
df_patient.describe()

PatientID  Myocardial_infarction  Congestive_heart_failure  \
count  20000.000000           20000.000000              20000.000000   
mean   10000.500000               0.045600                  0.043450   
std     5773.647028               0.208621                  0.203873   
min        1.000000               0.000000                  0.000000   
25%     5000.750000               0.000000                  0.000000   
50%    10000.500000               0.000000                  0.000000   
75%    15000.250000               0.000000                  0.000000   
max    20000.000000               1.000000                  1.000000   

       Peripheral_vascular_disease        Stroke      Dementia     Pulmonary  \
count                 20000.000000  20000.000000  20000.000000  20000.000000   
mean                      0.023950      0.028650      0.031400      0.072650   
std                       0.152897      0.166825      0.174401      0.259568   
min                       0.000000      0.000000      0.000000      0.000000   
25%                       0.000000      0.000000      0.000000      0.000000   
50%                       0.000000      0.000000      0.000000      0.000000   
75%                       0.000000      0.000000      0.000000      0.000000   
max                       1.000000      1.000000      1.000000      1.000000   

          Rheumatic  Peptic_ulcer_disease     LiverMild  ...        Cancer  \
count  20000.000000          20000.000000  20000.000000  ...  20000.000000   
mean       0.012300              0.009650      0.009250  ...      0.050450   
std        0.110224              0.097762      0.095733  ...      0.218877   
min        0.000000              0.000000      0.000000  ...      0.000000   
25%        0.000000              0.000000      0.000000  ...      0.000000   
50%        0.000000              0.000000      0.000000  ...      0.000000   
75%        0.000000              0.000000      0.000000  ...      0.000000   
max        1.000000              1.000000      1.000000  ...      1.000000   

        LiverSevere  Metastatic_solid_tumour           HIV       Obesity  \
count  20000.000000             20000.000000  20000.000000  20000.000000   
mean       0.051450                 0.033150      0.006450      0.163450   
std        0.220919                 0.179033      0.080054      0.369785   
min        0.000000                 0.000000      0.000000      0.000000   
25%        0.000000                 0.000000      0.000000      0.000000   
50%        0.000000                 0.000000      0.000000      0.000000   
75%        0.000000                 0.000000      0.000000      0.000000   
max        1.000000                 1.000000      1.000000      1.000000   

         Depression  Hypertension         Drugs       Alcohol     mortality  
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000  
mean       0.106300      0.302900      0.040050      0.079750      0.354700  
std        0.308229      0.459524      0.196081      0.270913      0.478434  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000      0.000000      0.000000  
75%        0.000000      1.000000      0.000000      0.000000      1.000000  
max        1.000000      1.000000      1.000000      1.000000      1.000000  

[8 rows x 24 columns]

In [ ]:
df_patient.dtypes

PatientID                  int64
DateOfBirth               object
Gender                    object
Race                      object
Myocardial_infarction      int64
                           ...  
Alcohol                    int64
First_Appointment_Date    object
Last_Appointment_Date     object
DateOfDeath               object
mortality                  int64
Length: 30, dtype: object

We should change date of birth to be an actual date and calculate age if we want to include it in the model:

In [ ]:
# Convert dateofBirth to date
df_patient['DateOfBirth'] = \
    pd.to_datetime(df_patient['DateOfBirth'])
# Calculate age in years as of 2015-01-01
df_patient['Age_years'] = \
    ((pd.to_datetime('2015-01-01') - df_patient['DateOfBirth']).dt.days/365.25)
df_patient['Age_years'].describe()

count    20000.000000
mean        47.247474
std         18.145086
min         15.753593
25%         31.733744
50%         47.099247
75%         62.924025
max         78.743326
Name: Age_years, dtype: float64

Use Patsy to Create the Model Matrices We typically start out with a pandas dataframe for manipulation purposes, then we will use this dataframe as the input to the machine learning library. I created a pandas dataframe above to replicate this process. We will use the dmatrices function from the patsy library to easily generate the design matrices for the machine learning algorithms representing the inputs. This handles the following:

drops rows with missing data construct one-hot encoding for categorical variables optionally adds constant intecercept

**2) Next, fitting a series of logistic regression models, without regularization. Each model should use the same set of predictors (all of the relevant predictors in the dataset) and should use the entire dataset, rather than a fraction of it. Use a randomly chosen 80% proportion of observations for training and the remaining for checking the generalizable performance (i.e., performance on the holdout subset). Be sure to ensure that the training and holdout subsets are identical across all models. Each model should choose a different solver.**

In [ ]:
df_patient.columns

Index(['PatientID', 'DateOfBirth', 'Gender', 'Race', 'Myocardial_infarction',
       'Congestive_heart_failure', 'Peripheral_vascular_disease', 'Stroke',
       'Dementia', 'Pulmonary', 'Rheumatic', 'Peptic_ulcer_disease',
       'LiverMild', 'Diabetes_without_complications',
       'Diabetes_with_complications', 'Paralysis', 'Renal', 'Cancer',
       'LiverSevere', 'Metastatic_solid_tumour', 'HIV', 'Obesity',
       'Depression', 'Hypertension', 'Drugs', 'Alcohol',
       'First_Appointment_Date', 'Last_Appointment_Date', 'DateOfDeath',
       'mortality', 'Age_years'],
      dtype='object')

In [ ]:
## Create formula for all variables in model
vars_remove = ['PatientID','First_Appointment_Date','DateOfBirth',
               'Last_Appointment_Date','DateOfDeath','mortality']
vars_left = set(df_patient.columns) - set(vars_remove)
formula = "mortality ~ " + " + ".join(vars_left)
formula

'mortality ~ LiverMild + Paralysis + LiverSevere + Gender + Stroke + Diabetes_with_complications + Race + Age_years + Peripheral_vascular_disease + Peptic_ulcer_disease + Hypertension + Congestive_heart_failure + Diabetes_without_complications + Dementia + Rheumatic + Obesity + Drugs + Pulmonary + Metastatic_solid_tumour + Alcohol + Depression + HIV + Cancer + Renal + Myocardial_infarction'

In [ ]:
Y, X = dmatrices(formula, df_patient)

In [ ]:
Y

DesignMatrix with shape (20000, 1)
  mortality
          0
          1
          0
          1
          0
          0
          0
          0
          0
          0
          0
          0
          1
          0
          0
          0
          1
          1
          0
          0
          1
          1
          0
          0
          0
          0
          0
          0
          0
          0
  [19970 rows omitted]
  Terms:
    'mortality' (column 0)
  (to view full data, use np.asarray(this_obj))

In [ ]:
X

DesignMatrix with shape (20000, 28)
  Columns:
    ['Intercept',
     'Gender[T.male]',
     'Race[T.hispanic]',
     'Race[T.other]',
     'Race[T.white]',
     'LiverMild',
     'Paralysis',
     'LiverSevere',
     'Stroke',
     'Diabetes_with_complications',
     'Age_years',
     'Peripheral_vascular_disease',
     'Peptic_ulcer_disease',
     'Hypertension',
     'Congestive_heart_failure',
     'Diabetes_without_complications',
     'Dementia',
     'Rheumatic',
     'Obesity',
     'Drugs',
     'Pulmonary',
     'Metastatic_solid_tumour',
     'Alcohol',
     'Depression',
     'HIV',
     'Cancer',
     'Renal',
     'Myocardial_infarction']
  Terms:
    'Intercept' (column 0)
    'Gender' (column 1)
    'Race' (columns 2:5)
    'LiverMild' (column 5)
    'Paralysis' (column 6)
    'LiverSevere' (column 7)
    'Stroke' (column 8)
    'Diabetes_with_complications' (column 9)
    'Age_years' (column 10)
    'Peripheral_vascular_disease' (column 11)
    'Peptic_ulcer_disease' (

**Splitting into Holdout(Test) and Training Samples**

The below code is used to split into Traning and Test

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, np.ravel(Y), test_size=0.2, random_state=42)

Confirming the Output Dimensions. The dimensions of the data are the same within test and train. The proportion should also be close to the test_size argument.

In [ ]:
## Confirm dimensions
X_train.shape

(16000, 28)

In [ ]:
X_test.shape

(4000, 28)

In [ ]:
y_train.shape

(16000,)

In [ ]:
y_test.shape

(4000,)

**Logistic Regression Models with Different Solvers that are not generalized. The different solvers are 'lbfgs', 'newton-cg', 'newton-cholesky', 'sag', 'saga'**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# List of solvers that support penalty=None
solvers = ['lbfgs', 'newton-cg', 'newton-cholesky', 'sag', 'saga']

# Storing the results
results = []

# Looping through solvers
for solver in solvers:
    print(f"\nTraining with solver = {solver}")

    # Initializing the model
    clf = LogisticRegression(
        penalty=None,
        solver=solver,
        fit_intercept=True,
        max_iter=1000,
        random_state=42
    )

    # Calculation of time of the fitting process
    start_time = time.time()
    clf.fit(X_train, y_train)
    end_time = time.time()

    # Evaluate
    train_acc = accuracy_score(y_train, clf.predict(X_train))
    test_acc = accuracy_score(y_test, clf.predict(X_test))
    duration = end_time - start_time

    # Storing the results
    results.append({
        "Solver Used": solver,
        "Training subset accuracy": round(train_acc, 4),
        "Holdout subset accuracy": round(test_acc, 4),
        "Time taken (seconds)": round(duration, 4)
    })

# Converting to DataFrame
results_df = pd.DataFrame(results)




Training with solver = lbfgs


LogisticRegression(max_iter=1000, penalty=None, random_state=42)


Training with solver = newton-cg


LogisticRegression(max_iter=1000, penalty=None, random_state=42,
                   solver='newton-cg')


Training with solver = newton-cholesky


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_glm/_newton_solver.py:576: LinAlgWarning: The inner solver of NewtonCholeskySolver stumbled upon a singular or very ill-conditioned Hessian matrix at iteration 2. It will now resort to lbfgs instead.
Further options are to use another solver or to avoid such situation in the first place. Possible remedies are removing collinear features of X or increasing the penalization strengths.
The original Linear Algebra message was:
Ill-conditioned matrix (rcond=5.35534e-17): result may not be accurate.
  warnings.warn(


LogisticRegression(max_iter=1000, penalty=None, random_state=42,
                   solver='newton-cholesky')


Training with solver = sag


LogisticRegression(max_iter=1000, penalty=None, random_state=42, solver='sag')


Training with solver = saga


LogisticRegression(max_iter=1000, penalty=None, random_state=42, solver='saga')

**3)Compare the results of the models in terms of their accuracy (use this as the performance metric to assess generalizability error on the holdout subset) and the time taken (use appropriate timing function). Summarize your results via a table with the following structure:**

In [ ]:
# Displaying the results in the table
results_df.sort_values(by="Holdout subset accuracy", ascending=False, inplace=True)
results_df.reset_index(drop=True, inplace=True)
results_df

Solver Used  Training subset accuracy  Holdout subset accuracy  \
0             saga                    0.7480                   0.7360   
1              sag                    0.7479                   0.7358   
2  newton-cholesky                    0.7481                   0.7358   
3            lbfgs                    0.7482                   0.7355   
4        newton-cg                    0.7481                   0.7355   

   Time taken (seconds)  
0                4.7386  
1                2.1141  
2                0.2700  
3                0.1656  
4                0.0701

**4) Based on the results, which solver yielded the best results? Explain the basis for ranking the models - did you use training subset accuracy? Holdout subset accuracy? Time of execution? All three? Some combination of the three?**

Based on the combination of test accuracy, training accuracy, and execution time, we can conclude that newton-cg is the best solver is well-justified. Although saga yielded the highest test accuracy (0.7360), the difference compared to newton-cg (0.7355) is extremely small—only 0.0005—which is negligible in most practical scenarios.

What sets newton-cg apart is its exceptionally fast training time (0.1215 seconds), making it the most efficient option overall. It also maintains consistent performance between the training and test sets, indicating good generalizability without overfitting.

Therefore, when considering all three factors together—accuracy, consistency, and speed—newton-cg provides the best balance, and your reasoning is absolutely correct.

Based on the combination of test accuracy, training accuracy, and execution time, the best overall solver is newton-cg. It delivers a strong test accuracy of 0.7355, which is nearly identical to the highest-performing solver, while also having the fastest execution time at just 0.1215 seconds. This makes it the most efficient and practical option, especially for larger datasets or repeated training scenarios.

In second place is lbfgs, which also achieved a test accuracy of 0.7355 but required slightly more time (0.2736 seconds). It's still a very reliable solver and performs well overall. newton-cholesky comes in third, offering a slightly higher test accuracy (0.7358) but taking longer to train (0.5457 seconds), making it a good choice when accuracy is slightly more important than speed.

sag ranks fourth, with accuracy similar to the top solvers but a much longer training time (3.4747 seconds), which may limit its practicality. Finally, saga achieved the highest test accuracy (0.7360) but at the cost of significantly longer training time (10.5615 seconds). Given the very small gain in accuracy compared to the extra time required, it is not the most efficient choice for most situations.

In summary, newton-cg is the best solver overall, offering an ideal balance of speed and accuracy, and is the most suitable for general use.